src/main/java/org/broadinstitute/hellbender/tools/HtsgetReader.java

```diff
@@ -0,0 +1,441 @@
+
+    @Override
+    public void onTraversalStart() {
+        if (this.readerThreads > 1) {
+            final ThreadFactory threadFactory = new ThreadFactoryBuilder()
+                    .setNameFormat("htsgetReader-thread-%d")
+                    .setDaemon(true)
+                    .build();
+            this.executorService = Executors.newFixedThreadPool(readerThreads, threadFactory);
+        }
+        this.client = HttpClientBuilder.create()
+            .setConnectionManager(new PoolingHttpClientConnectionManager())
+            // throw IOException on 5xx error codes so request is retried
+            .addInterceptorLast(new HttpResponseInterceptor() {
+                @Override
+                public void process(HttpResponse resp, HttpContext ctxt) throws HttpException, IOException {
+                    int statusCode = resp.getStatusLine().getStatusCode();
+                    if (500 <= statusCode && statusCode < 600) {
+                        throw new IOException("Retry on 5xx error code");
+                    }
+                }
+            })
+            .build();
+    }
+
+    @Override
+    public void onShutdown() {
+        try {
+            this.client.close();
+        } catch (IOException e) {
+            logger.warn("Exception while closing httpClient");
+        }
+    }
+
+    /**
+     * Validates that the user query obeys htsget spec 
+     */
+    private void validateRequest(HtsgetRequest req) {
+        if (req.getDataClass() != null && req.getDataClass() == HtsgetClass.header && (
+            req.getReferenceName() != null ||
+            req.getStart() != null ||
+            req.getEnd() != null ||
+            ! req.getFields().isEmpty() ||
+            ! req.getTags().isEmpty() ||
+            ! req.getNoTags().isEmpty())) {
+                throw new UserException("Invalid request: no query parameters except `format` may be specified when class=header");
+        }
+        if (req.getStart() != null && req.getStart() < 0) {
+            throw new UserException("Invalid request: start must be non-negative");
+        }
+        if (req.getEnd() != null && req.getEnd() < 0) {
+            throw new UserException("Invalid request: end must be non-negative");
+        }
+        if (req.getStart() != null && req.getEnd() != null && req.getStart() > req.getEnd()) {
+            throw new UserException("Invalid request: end may not be greater than start");
+        }
+        String intersections = req.getTags().stream()
+            .filter(req.getNoTags()::contains)
+            .collect(Collectors.joining(", "));
+        if (! intersections.isEmpty()) {
+            throw new UserException("Invalid request: tags and notags overlap in the following fields: " + intersections);
+        }
+    }
+
+    /** 
+     * Downloads data blocks provided by response to outputFile in serial
+    */
+    private void getData(HtsgetResponse response) throws IOException {
```

Comment:
Changed by moving the code for getting an OutputStream to the Block class. I decided against splitting Block into two classes because this would complicate deserializing with Jackson more than it would be worth, in my opinion